### PR TITLE
Improve scoring and pass computations

### PR DESCRIPTION
**Pull request by alexander-de-ranitz**
### Comments for the reviewer
For more information regarding the changes to the scoring algorithms, see [this document](https://docs.google.com/document/d/1slbu_LbTG13PoeqacebppXwy4W5sBZnTIz9mHPmfERg/edit)

###### Code Quality
- [x] Is the code is understandable and easy to read
- [x] Changes to the code comply with set clang-format rules
- [x] No use of manual memory control (e.g new/malloc/colloc etc)
- [x] Are (only) smart pointers used?

###### Testing
- [ ] All tests are passing.
- [x] I _added new / changed existing_ tests to reflect code changes (state why not otherwise!)
- [x] I tested my changes manually (Describe how, to what extent etc.)

###### Commit Messages
- [x] Commit message is saying what has been changed, **why** it was changed? Remember other developers might not know
  what the problem you are fixing was. Note also negative _decision_ (e.g., why did you not do particular thing)
  **TLDR: Commit message are comprehensive**
- [x] Commit messages follows the rules of https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: include/roboteam_ai/stp/computations/PassComputations.h [hunk 1]
```diff
@@ -28,6 +28,7 @@ class PassComputations {
      * @param ballLocation location of the ball
```
Comment (Florisheinen1) [inline include/roboteam_ai/stp/computations/PassComputations.h:28]: Could you write the function description more clear? It looks to me like the function wants to pass to the robots instead of other points in the field

File: src/stp/computations/PassComputations.cpp [hunk 2]
```diff
@@ -17,28 +17,27 @@ gen::ScoredPosition PassComputations::calculatePassLocation(Vector2 ballLocation
...
+            if (!FieldComputations::pointIsValidPosition(field, point, 2 * control_constants::ROBOT_RADIUS)) continue;
+            if (PositionScoring::scorePosition(point, gen::LineOfSight, field, world).score < 10) continue;  // Need minimum LoS to be a valid pass. Avoid passing into enemies
             for (auto robotLocation : robotLocations) {
-                if (FieldComputations::pointIsValidPosition(field, point, 2 * control_constants::ROBOT_RADIUS) &&
-                    calculateRobotTravelTime(robotLocation, point) < calculateBallTravelTime(ballLocation, passerLocation, point)) {
-                    if (PositionScoring::scorePosition(point, gen::LineOfSight, field, world).score < 10) {  // Need minimum LoS to be a valid pass. Avoid passing into enemies
-                        continue;
-                    }
-                    auto scoredPosition = PositionScoring::scorePosition(point, profile, field, world);
+                if (calculateRobotTravelTime(robotLocation, point) < calculateBallTravelTime(ballLocation, passerLocation, point)) {
```
Comment (Florisheinen1) [inline src/stp/computations/PassComputations.cpp:38]: Could you explain in comments what this piece of code is doing?

File: src/stp/computations/PassComputations.cpp [hunk 3]
```diff
@@ -17,28 +17,27 @@ gen::ScoredPosition PassComputations::calculatePassLocation(Vector2 ballLocation
...
         return bestPassLocation;
     }
     // TODO: create a better passingGrid (more points and/or better spread)
-    auto passGrid = Grid(-gridLength / 2, -gridWidth / 2, gridWidth, gridLength, numPoints, numPoints);  // 9 points spread over the whole field
+    auto passGrid = Grid(-gridLength / 2, -gridWidth / 2, gridWidth, gridLength, numPoints, numPoints);  // 81 points spread over the whole field
     for (auto& pointVector : passGrid.getPoints()) {
         for (auto& point : pointVector) {
-            if (point.dist(ballLocation) < 1) continue;  // Do not pass less than 1m far- this can be dribbled
+            if (point.dist(ballLocation) < 1) continue; // Do not pass less than 1m far- this can be dribbled
+            if (!FieldComputations::pointIsValidPosition(field, point, 2 * control_constants::ROBOT_RADIUS)) continue;
```
Comment (Florisheinen1) [inline src/stp/computations/PassComputations.cpp:34]: This way it is impossible to pass to the keeper. Should this be changed?

File: src/stp/computations/PassComputations.cpp [hunk 4]
```diff
@@ -17,28 +17,27 @@ gen::ScoredPosition PassComputations::calculatePassLocation(Vector2 ballLocation
...
         RTT_WARNING("No possible receivers!")
         return bestPassLocation;
     }
     // TODO: create a better passingGrid (more points and/or better spread)
-    auto passGrid = Grid(-gridLength / 2, -gridWidth / 2, gridWidth, gridLength, numPoints, numPoints);  // 9 points spread over the whole field
+    auto passGrid = Grid(-gridLength / 2, -gridWidth / 2, gridWidth, gridLength, numPoints, numPoints);  // 81 points spread over the whole field
     for (auto& pointVector : passGrid.getPoints()) {
         for (auto& point : pointVector) {
-            if (point.dist(ballLocation) < 1) continue;  // Do not pass less than 1m far- this can be dribbled
+            if (point.dist(ballLocation) < 1) continue; // Do not pass less than 1m far- this can be dribbled
```
Comment (Florisheinen1) [inline src/stp/computations/PassComputations.cpp]: Could you define this `1` as a constexpr variable 'MINIMUM_PASS_DISTANCE' for clarity?

File: src/stp/computations/PassComputations.cpp [hunk 5]
```diff
@@ -17,28 +17,27 @@ gen::ScoredPosition PassComputations::calculatePassLocation(Vector2 ballLocation
...
+            if (!FieldComputations::pointIsValidPosition(field, point, 2 * control_constants::ROBOT_RADIUS)) continue;
+            if (PositionScoring::scorePosition(point, gen::LineOfSight, field, world).score < 10) continue;  // Need minimum LoS to be a valid pass. Avoid passing into enemies
```
Comment (Florisheinen1) [inline src/stp/computations/PassComputations.cpp]: And perhaps also put this minimum pass score in a constexpr variable

File: include/roboteam_ai/stp/computations/PositionScoring.h [hunk 6]
```diff
@@ -19,7 +19,7 @@ class PositionScoring {
      * @param scores ref to struct linked to that pos
```
Comment (Florisheinen1) [inline include/roboteam_ai/stp/computations/PositionScoring.h:20]: Please add the new Field parameter in the doxygen comments

File: src/stp/evaluations/position/BlockingEvaluation.cpp [hunk 7]
```diff
@@ -12,7 +12,7 @@ uint8_t BlockingEvaluation::metricCheck(double pointDistance, std::vector<double
     double maxAngle = M_PI / 6;  // 30 degrees, max angle considered for blocking
```
Comment (Florisheinen1) [inline src/stp/evaluations/position/BlockingEvaluation.cpp]: Please make this constexpr so this will be calculated during compilation (in DEBUG mode. Release mode will most likely already optimize it)

File: src/stp/evaluations/position/LineOfSightEvaluation.cpp [hunk 8]
```diff
@@ -10,29 +10,26 @@
...
+     *        -_                                       /
+     *           -_                                   /
      *              -_                               /
      */
 
-    double evalScore = 1;           // Default score (= perfect line of sight)
-    double innerAngle = M_PI / 36;  // 5 degrees
-    double outerAngle = M_PI / 6;   // 30 degrees
+    double evalScore = 1;          // Default score (= perfect line of sight)
+    double outerAngle = M_PI / 6;  // 30 degrees- enemies outside this angle are not considered
```
Comment (Florisheinen1) [inline src/stp/evaluations/position/LineOfSightEvaluation.cpp]: Please make this constexpr too

File: include/roboteam_ai/stp/computations/PositionScoring.h [hunk 9]
```diff
@@ -19,7 +19,7 @@ class PositionScoring {
...
      * @return score value
      */
-    static double determineOpenScore(Vector2 &point, const world::World *world, gen::PositionScores &scores);
+    static double determineOpenScore(Vector2 &point, const rtt::world::Field &field, const world::World *world, gen::PositionScores &scores);
```
Comment (Florisheinen1) [inline include/roboteam_ai/stp/computations/PositionScoring.h:23]: Why add field as a parameter when it can also be obtained through world?

**Review by Florisheinen1 — COMMENTED** [review]
Great work on improving these scorings, and thanks for providing the document with extra explanation and the nice visualizations.
I ran AI with this, and everything still seems to work, so it is already merge worthy. If you can implement some of the changes I requested, I will merge it

File: src/stp/computations/PassComputations.cpp (hunk 3 above)
Comment (alexander-de-ranitz) [reply src/stp/computations/PassComputations.cpp:34]: That is intentional as a safety measure :) Receiving the pass is not 100% consistent so it would be a shame if we score own goals hahah

File: include/roboteam_ai/stp/computations/PositionScoring.h (hunk 9 above)
Comment (alexander-de-ranitz) [reply include/roboteam_ai/stp/computations/PositionScoring.h:23]: Speed. Passing it to the function as a separate argument speeds up the whole function by a factor of around 30 compared to extracting field from world on each call.

**Review by Florisheinen1 — APPROVED** [review]
LGTM
